src/main/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryTaskLauncher.java

```diff
@@ -0,0 +1,620 @@
+
+    private final CloudFoundryClient client;
+
+    private final CloudFoundryOperations operations;
+
+    private final CloudFoundryDeployerProperties properties;
+
+    public static final String SERVICES_PROPERTY_KEY = "spring.cloud.deployer.cloudfoundry.defaults.services";
+
+    public static final String MEMORY_PROPERTY_KEY = "spring.cloud.deployer.cloudfoundry.defaults.memory";
+
+    public static final String DISK_PROPERTY_KEY = "spring.cloud.deployer.cloudfoundry.defaults.disk";
+
+    private long timeout = 30;
+
+    public CloudFoundryTaskLauncher(CloudFoundryClient client, CloudFoundryOperations operations, CloudFoundryDeployerProperties properties) {
+        this.client = client;
+        this.operations = operations;
+        this.properties = properties;
+    }
+
+    /**
+     * @param timeout timeout in seconds for blocking events (status and launch).
+     */
+    public void setTimeout(long timeout) {
+        this.timeout = timeout;
+    }
+
+    /**
+     * Setup a reactor pipeline to cancel a running task.
+     *
+     * @param id the task's id to be cancled as returned from the {@link TaskLauncher#launch(AppDeploymentRequest)}
+     */
+    @Override
+    public void cancel(String id) {
+
+        asyncCancel(id).subscribe();
+    }
+
+    /**
+     * Set up a reactor pipeline to launch a task. Before launch, check if it exists. If not, deploy. Then launch.
+     *
+     * @param request description of the application to be launched
+     * @return name of the launched task, returned without waiting for reactor pipeline to complete
+     */
+    @Override
+    public String launch(AppDeploymentRequest request) {
+        return asyncLaunch(request).block(Duration.ofSeconds(this.timeout));
+    }
+
+    /**
+     * Lookup the current status based on task id.
+     *
+     * @param id taskId as returned from the {@link TaskLauncher#launch(AppDeploymentRequest)}
+     * @return the current task status
+     */
+    @Override
+    public TaskStatus status(String id) {
+
+        return asyncStatus(id).block(Duration.ofSeconds(this.timeout));
+    }
+
+    /**
+     * Issues an async request to cancel a running task
+     *
+     * @param id the id of the task to be canceled
+     * @return A Mono that will return the results of the cancel request.
+     */
+    protected Mono<CancelTaskResponse> asyncCancel(String id) {
+
+        return client.tasks().cancel(CancelTaskRequest.builder()
+            .taskId(id)
+            .build());
+    }
+
+    /**
+     * Issues an async request to launch a task
+     *
+     * @param request an {@link AppDeploymentRequest} containing the details of the task to be launched
+     * @return A Mono that will return the id of the task launched
+     */
+    protected Mono<String> asyncLaunch(AppDeploymentRequest request) {
+        return client.applicationsV3().list(ListApplicationsRequest.builder()
+                .name(request.getDefinition().getName())
+                .page(1)
```

Comment:
Probably not wise to assume that there's only a single page of applications when you list them.  There's now a [`PaginationUtils#requestClientV3Resources](http://docs.cloudfoundry.com/cf-java-client/api/latest-snapshot/cloudfoundry-util/org/cloudfoundry/util/PaginationUtils.html#requestClientV3Resources-java.util.function.Function-) that you can use to help with paginated requests and responses.